0000-error-values-proposal.rst

```diff
@@ -0,0 +1,282 @@
+      | ...
+
+    warningMessage :: GHCWarning -> ErrMsg
+
+    type ErrorMessages   = Bag GHCError
+    type WarningMessages = Bag GHCWarning
+    type Messages        = (WarningMessages, ErrorMessages) -- as before
+
+(The alternative being to just store ``GHCError`` values in both bags and
+augment ``GHCError`` with a constructor dedicated to warnings.)
+
+Finally, we would have to update some error reporting infrastructure
+to take ``GHCError`` values as arguments instead of ``ErrMsg``. That is
+the point at which the actual rendering of error messages would happen,
+under this proposal, right before calling the code that logs the said errors.
+
+A consequence of this is that the ``Messages`` type that GHC API users
+consume would now carry error and warning **values** that they can render
+but also inspect, without parsing. A lot of the work would be about
+actually moving all the error rendering code away from where we create
+errors, and defining suitable types that carry the data around until
+it is time to report the errors to the user.
+
+Effect and Interactions
+-----------------------
+
+By turning errors into proper values, tooling authors would be able to
+get rid of their error parsing code and finally be able to concisely
+inspect, render or "customize" error messages. This is the main attraction of
+this proposal. However, we list below a few compelling applications that are at
+best very cumbersome to write with the current error documents and which would
+be made a lot easier if the current proposal is accepted.
+
+* An IDE tool might want to gather the bindings and their types, as listed by
+  GHC in some type errors, to allow editors to display those in a tooltip or
+  minibuffer, helping the developer figure out the right combination of those
+  bindings by visually placing this information next to the code.
+
+* An IDE tool might use the suggestions that GHC would embed in error
+  values to present automated refactoring options to the user (e.g enabling
+  language extensions, importing a module, fixing a typo in the spelling of
+  a name).
+
+* Display squiggles of different colors depending on the nature of the error
+  (e.g "not in scope" errors in some color, "couldn't match" errors in another,
+  and so on, or perhaps discriminating in larger groups). This could help
+  Haskell developers focus on one of those groups (e.g typos in names,
+  missing imports) before proceeding with the more subtle type errors.
+
+* An IDE tool might ask GHC to defer expensive analyses typically done
+  during error message construction (e.g. `computing valid hole fits
+  <https://gitlab.haskell.org/ghc/ghc/issues/16875#note_210045>`_) and instead
+  query GHC for the analysis result asynchronously (or even only when
+  requested by the user), shrinking the edit/typechecking iteration time.
+
+Costs and drawbacks
+-------------------
+
+The import cycles induced by this architecture and the ``.hs-boot`` files that
```

Comment:
If you agree with my analysis that only one hs-boot file will be necessary, revise this bit.